### PR TITLE
fea(#507): support assign `--post-renderer` flags ， helmDefaults  config ， release config when use helm v3

### PR DESCRIPTION
**Pull request by tanguofu**
Signed-off-by: guofutan <guofutan@tencent.com>

This implement  reuse the args in helmfile args and helmDefaults config of helmfile.yaml， these PR to maske sure the less change to helmfile code， but make sure the forward compatbility of the helmfile.yaml syntx

### CODE REVIEW

File: pkg/helmexec/exec.go [hunk 1]
```diff
@@ -131,7 +132,19 @@ func New(helmBinary string, enableLiveOutput bool, logger *zap.SugaredLogger, ku
 }
 
 func (helm *execer) SetExtraArgs(args ...string) {
-	helm.extra = args
+	var extraArgs []string
+	
+	// reset the postRenderers and filter the post-renderer args from args and put into helm.postRenderers
+	helm.postRenderers = nil
+	for _, arg := range args {
+		if strings.HasPrefix(arg, "--post-renderer=") || strings.HasPrefix(arg, "--post-renderer-args=") {
```
Comment (yxxhero) [inline pkg/helmexec/exec.go:140]: when use a format like the below:
```
--post-renderer  value
```

what will happen?

**Comment by yxxhero** [comment]
@tanguofu Need most units and e2e tests.

File: pkg/helmexec/exec.go (hunk 1 above)
Comment (tanguofu) [reply pkg/helmexec/exec.go:140]: this format will be deal with [analyzeArgs](https://github.com/helmfile/helmfile/blob/main/pkg/argparser/args.go#L57) ， and keep the args  order then past into helm command：
such as :
```
helmfile -e ksp -f ti-k8s/helmfile.yaml -l chart=mysql  --args="--post-renderer=rewrite-ksp.sh"
```

the helm command is : 
```
 helm template mysql mysql --namespace ti-inf --values /tmp/helmfile3923818193/ti-inf-mysql-values-674855cccb --values /tmp/helmfile1141692238/ti-inf-mysql-values-9b4dd874f --values /tmp/helmfile2042720088/ti-inf-mysql-values-5448c6fcbd --values /tmp/helmfile1303814841/ti-inf-acp-mysql-values-776bd65866 --post-renderer rewrite-ksp.sh --debug
```
so these will also work well.

Comment (yxxhero) [reply pkg/helmexec/exec.go:140]: @tanguofu 
```
if strings.HasPrefix(arg, "--post-renderer=") || strings.HasPrefix(arg, "--post-renderer-args=") {
```

will `--post-renderer=vaule` `--post-renderer value` all be handled?

**Review by mumoshu — CHANGES_REQUESTED** [review]
Hey! I just took a glance at your awesome work. One question- Although you mentioned helmDefaults in the PR title, the actual change does not contain anything related to helmDefaults.
Are you also willing to add the post renderer support to helmDefaults too? That would be great if so.

**Comment by tanguofu** [comment]
> @tanguofu BTW. please fix ci issue and add tests.

thanks  to @yxxhero  for careful  CR!

i add the process when the args like `--post-renderer=”xxx xx.sh“`
and  add unit test at `func Test_SetExtraArgs(t *testing.T)`  fix the `golangci-lint run ` errors .
may  u take a review again ? 
thanks very much.

**Comment by tanguofu** [comment]
> Hey! I just took a glance at your awesome work. One question- Although you mentioned helmDefaults in the PR title, the actual change does not contain anything related to helmDefaults. Are you also willing to add the post renderer support to helmDefaults too? That would be great if so.

i think it is not a good idea to add a new config `post-renderer` inito helmDefaults after me write the PR.  instead i thinsk it is better to reuse  `args` config in helmDefaults; becasue:
1. add a new config is harmful for previous versions helmfile.yaml (ie maybe some tools which parse helmfile.yaml the  suppose that  is invariable）
2. in code implement there will be more files and codes change to helmfile codebase. which also take more cost to test。
So resue the args  also make these command works :
```
helmfile -e ksp -f ti-k8s/helmfile.yaml -l chart=mysql  --args="--post-renderer rewrite-ksp.sh"
```
this will make this feature more easy to use.

**Comment by mumoshu** [comment]
@tanguofu Helmfile is designed for use when you want to manage a set of releases declaratively. If you need a specific post-renderer to make your release(s) get deployed properly, I consider the post-renderer a part of the deployment, which means it should be covered in helmfile.yaml.

> add a new config is harmful for previous versions helmfile.yaml (ie maybe some tools which parse helmfile.yaml the suppose that is invariable）

We don't retain forward compatibility in Helmfile, like other projects within the K8s community. We do maintain backward compatibility. In the context of backward compatibility, adding new fields is fine.

**Comment by tanguofu** [comment]
> @tanguofu Helmfile is designed for use when you want to manage a set of releases declaratively. If you need a specific post-renderer to make your release(s) get deployed properly, I consider the post-renderer a part of the deployment, which means it should be covered in helmfile.yaml.
> 
> > add a new config is harmful for previous versions helmfile.yaml (ie maybe some tools which parse helmfile.yaml the suppose that is invariable）
> 
> We don't retain forward compatibility in Helmfile. I think that's generally the same stance as other projects within the K8s community. We do maintain backward compatibility. In the context of backward compatibility, adding new fields is fine.

well let me add the `post-renderer` configs in helmDefaults of `helmfile.yaml`.

**Comment by tanguofu** [comment]
> @tanguofu Helmfile is designed for use when you want to manage a set of releases declaratively. If you need a specific post-renderer to make your release(s) get deployed properly, I consider the post-renderer a part of the deployment, which means it should be covered in helmfile.yaml.
> 
> > add a new config is harmful for previous versions helmfile.yaml (ie maybe some tools which parse helmfile.yaml the suppose that is invariable）
> 
> We don't retain forward compatibility in Helmfile. I think that's generally the same stance as other projects within the K8s community. We do maintain backward compatibility. In the context of backward compatibility, adding new fields is fine.
2
the `postRenderer` and `postRendererArgs` were both add in `helmDafult` of `helmfile.yaml`, please @mumoshu  review again

many thanks !

File: pkg/argparser/args.go [hunk 2]
```diff
@@ -112,6 +112,13 @@ func GetArgs(args string, state *state.HelmState) []string {
 		}
 	}
 
+	if len(state.HelmDefaults.PostRenderer) > 0 {
+		argArr = append(argArr, fmt.Sprintf("--post-renderer=\"%s\"", state.HelmDefaults.PostRenderer))
+	}
+	for _, arg := range state.HelmDefaults.PostRendererArgs {
+		argArr = append(argArr, fmt.Sprintf("--post-renderer-args=\"%s\"", arg))
+	}
+
```
Comment (mumoshu) [inline pkg/argparser/args.go:121]: I believe this is not the right place to add this code- The helmDefaults args, the under pinning of the `--args` flag of helmfile, has known to be not well-designed, and being removed in Helmfile v1. The main difficulty is that Helmfile has no way to know which args specified in the helmDefaults.args to be passed to which helm sub-commands (install, uninstall, list, diff, etc) a helmfile command is going to run.

More importantly, the post renderer should be passed down to `chartify` when you're going to use a directory of raw k8s manifests as a chart, or a kustomize config as a chart, or you're using some chartify-provided features like ad-hoc dependencies and ad-hoc patching(jsonPatches and stragicMergePatches, forceNamespace, and so on).

Still curious- did this actually work on your machine? 🤔

Comment (tanguofu) [reply pkg/argparser/args.go:121]: well,   the `--args` flag of helmfile is easy to use indeed, such as:
```
helmfile -e ksp -f ti-k8s/helmfile.yaml -l chart=mysql  --args="--post-renderer=`pwd`rewrite-ksp.sh"
```
as the `post renderer` accept a abs path of exec script, this works well in our project; 

though it is more `chartify` to special each flags to parse the flags to helm exactly； but based on what I know， it is need to refactor lots of  code,  may be change the `helmexec.Interface` and add some set methods
？  and  invoke at every  place of  `helmexec.SetExtraArgs` ?

So what is the good way to implement this feature？  may give some Guidelines please?   

thanks.

Comment (mumoshu) [reply pkg/argparser/args.go:121]: > this works well in our project;

Yep. But it's not guaranteed to work. We don't have a complete design doc and tests to make sure your usage of "--args" works. My understanding was that Helmfile must have been implemented to pass the args to every helm sub-commands, regardless of if the sub-command supports the set of args or not.

Comment (mumoshu) [reply pkg/argparser/args.go:121]: > the good way to implement this feature

I'd recommend following the same pattern that is used for other Helmfile flags and fields. Propagate Helmfile flags to helmDefaults, inherit helmDefaults to releases[] fields if necessary, and finally compose chartify parameters and helm flags from the final releases[] entry fields.

Comment (nautnatic) [reply pkg/argparser/args.go:121]: @tanguofu Can you implement the change suggested by @mumoshu? The solution with the possibility to set the post-rendering through flags, helmDefaults or per release sounds perfect. Missing Post-rendering is the only thing stopping me from switching to Helmfile, so I'd be very grateful. Keep up the good work, guys!

**Comment by yxxhero** [comment]
@tanguofu any updates?

**Comment by sherifkayad** [comment]
@mumoshu any updates on this one?

File: pkg/argparser/args.go (hunk 2 above)
Comment (tanguofu) [reply pkg/argparser/args.go:121]: > ok, i will  implement the change, please just wait a moment.

**Comment by KR411-prog** [comment]
This is very useful feature. Will this be merged soon?

**Comment by yxxhero** [comment]
@tanguofu waiting for your updates. thanks very much.

**Comment by mumoshu** [comment]
@sherifkayad @KR411-prog This is an open-source project driven by volunteers. Instead of a storm of "any updates / eta?" messages that force anyone to work on it for you for free, you and we should all be encouraged to contribute via code or test results that help the work to be done, if you need it!

**Comment by sherifkayad** [comment]
@mumoshu That's an interesting take on questions / issues placed to you projects .. in years of using / trying (as much as I can to enhance projects that I find doing something useful, that's really the first time I get such an unprofessional response from a maintainer .. anyways let's just get a few things clear here:
- I do apologize if my question was interpretted as "pressure" on people to do stuff ..
- When you say "me benfiting" or "me using" or whatever personal points, I do believe you are wrong! .. Many open source users will benefit from that not only "me"!
- Someone already contributed a PR that was up for review and I just wanted to see if it got in the ice box or if it was still worked on / prioritized
- It's absolutely your right as the maintainer to simply say: "sorry folks this is not a priority for us" .. instead of just storming out at people for asking about something they wanted!
- On why not simply contributing, well, you have to understand that not everyone that's using stuff, can really contribute! .. One dumb simple reason is e.g. know-how .. I for instance am too far away from being able to write Golang in a way that will satisfy even the least standards .. so basically I wasn't busy making money out of your open source project!

**Comment by yxxhero** [comment]
@tanguofu Thanks very much.

**Comment by mumoshu** [comment]
@sherifkayad From a professional volunteer opensource maintainer who usually works in mornings, nights and Sundays for everyone in the world for almost free and kept it for years, and who gets asked "What's ETA of something" and "Is there any update on this" thing so often, it's my responsibility to clearly state my stance, to avoid us maintainers and opensource contributors from burning out. I take that's one of my jobs as the primary maintainer of this project. If I didn't do it, who does it? Who saves us from burning out?

Also, you can always say something like "cheers!" or prefix your message with "Thanks a lot for all your efforts here! But ..." or something like that. Isn't that what we usually do in our real lives..? At least that's how I treat the maintainer or the contributor outside of this project, where I'm just an user of an opensource project.

All that said,

> It's absolutely your right as the maintainer to simply say: "sorry folks this is not a priority for us" .. instead of just storming out at people for asking about something they wanted!

I have something to learn from this 😉 Thank you for your feedback. But again, I hope every open-source user is more thoughtful about who's volunteering unpaid hours to respond to all your never-ending requests.

Sorry if this all sounded very "unprofessional". But that's how I continued this "job" for years.

Lastly, I tend to just put a 👍 on the pull request when I absolutely need something to be merged in a random repo, so that you can express that you need it without inviting more pressuring messages to the pull request. Just my two cents.

**Comment by mumoshu** [comment]
@sherifkayad Don't get me wrong. I do this to everyone who does the same. I'm not blaming you. It's not your fault at all if you have no time to contribute code or anything. It's due to the broken open-source economy. But still, I have to protect folks from burning out today, and, well, what you call "storming" turned out to be the only way that worked for me, unfortunately.

**Comment by mumoshu** [comment]
Well sorry for going too out-of-topic.

@tanguofu Thanks for the update! I've just reviewed your changes and almost everything looks great! One thing I have noticed is that you might need to add the post-renderer support also to our helm-diff wrapper code that is invoked on helmfile-diff and helmfile-apply. Let me see if I can add the support myself.

**Comment by yxxhero** [comment]
@mumoshu I support you.

**Comment by mumoshu** [comment]
Let me try if I can distill my [manual test procedure](https://github.com/helmfile/helmfile/pull/510#issuecomment-1346403226) into a new E2E test-case to keep it working.
Do you have anything to add other than that? @yxxhero

**Comment by yxxhero** [comment]
@mumoshu I think all is good.

**Comment by yxxhero** [comment]
@tanguofu maybe need some docs.

**Comment by yxxhero** [comment]
@mumoshu seem like we should add helm-diff when e2e.

**Comment by yxxhero** [comment]
@tanguofu @mumoshu maybe it require a k8s cluster to pass the e2e test.

**Comment by mumoshu** [comment]
@yxxhero Ah true! I didn't realize it until now.. Shall I revert [a3e8c04](https://github.com/helmfile/helmfile/pull/510/commits/a3e8c0426ceae5b600807580db2a54e991407a9b) so that we can merge this anyway?

**Comment by mumoshu** [comment]
Oops I pushed something to main by mistake and that closed this pull request... Let me fix it and resubmit this.

**Comment by mumoshu** [comment]
Okay we already disabled the ability to force-push `main`. Let's go ahead. Forget about the dirty commits and make the necessary changes in follow-up PRs 🙏

**Comment by mumoshu** [comment]
@tanguofu Thanks again for your contribution! We'd appreciate it if you could help in writing documentation about this feature when you have some time.

**Pull request merged**